### PR TITLE
add setting for aria-live assertive alert for ghost text

### LINKED ISSUE #167348: [Accessibility] div.monaco-tokenized-source requires aria-live="assertive"
CC @isidorn and @meganrogge

## Current Issue

1. Load NVDA or any screen reader.

1. Launch VSCode.

1. enable copilot extension.

1. Create any ts or js file.

1. Try to trigger copilot-powered inline suggestion (`Alt+\`).

1. See (hear) if suggestion pop-up is automatically read: it's not unless you press `Ctrl+K, Ctrl+I` to explicitly move your focus to the tooltip area.

* Notes: This is because `"div.monaco-tokenized-source"` does not have `aria-live="assertive"` to announce the dymanic content to screen reader.

* Notes: copilot is just one of the examples. Any extensions or widgets that depend `"div.monaco-tokenized-source"` have the same issue.

## Expected Behavior

The suggested dynamic content needs to be automatically announced to screen reader users.

* Note: I am willing to fix this issue and test with my screen reader.

## System Info

Version: 1.74.0-insider (system setup)
Commit: d0e44be43287acadb13f7c2808332e00fe52568d
Date: 2022-11-25T08:00:08.922Z
Electron: 19.1.6
Chromium: 102.0.5005.167
Node.js: 16.14.2
V8: 10.2.154.15-electron.0
OS: Windows_NT x64 10.0.22621
Sandboxed: Yes

### PR DESCRIPTION
**Pull request by meganrogge**
fixes https://github.com/microsoft/vscode/issues/167348

@jooyoungseo, since you last tested, I've:
- removed the debounce of the function
- added a setting `editor.ariaAssertiveInlineSuggestion` which is false by default
- removed the text `Suggestion: ` from the alert so that only the suggested text is read. There should be an audio cue played to indicate the suggestion shows up then you should receive the alert with the content it suggests.

Could you pls test and let us know if it's still what you expect?

### CODE REVIEW

**Comment by jooyoungseo** [comment]
@meganrogge -- thanks for doing this! I don't hear any audio cues when suggestion is available. The aria-live works well though.

I configured the following:

``` json
    "editor.ariaAssertiveInlineSuggestion": true,
    "audioCues.debouncePositionChanges": true
```

What is `audioCues.debouncePositionChanges` exactly for? I don't see this audio in the audio help menu, either.

**Comment by meganrogge** [comment]
>I don't hear any audio cues when suggestion is available.

What do you have `audioCues.lineHasInlineSuggestion` set to?

>     "audioCues.debouncePositionChanges": true

You'll want to set this to false so you hear audio cues when you move your cursor and are on a word with an error for example. There were some ppl who reported issues related to hearing too many audio cues, which is why that is a setting and configurable. 

Here were the issues that led us to have that setting:
https://github.com/microsoft/vscode/issues/142121
https://github.com/microsoft/vscode/issues/143727
https://github.com/microsoft/vscode/issues/141444

**Comment by jooyoungseo** [comment]
@meganrogge -- Thanks for your detailed explanation! I have already turned it on: `{    "audioCues.lineHasInlineSuggestion": "on"}`

Isn't inline suggestion audio cues only for the inline hint such as copilot? this is not playing against normal auto-suggestion.

**Comment by meganrogge** [comment]
That is correct, it will only play when there's an inline hint such as copilot by design.

**Comment by jooyoungseo** [comment]
@meganrogge

> removed the text Suggestion:  from the alert so that only the suggested text is read.  

Since we have taken this out, I think we need the audio cues to audibly indicate the suggestion popup.

**Comment by meganrogge** [comment]
When I test, here's what I see/hear:

- inline suggestion appears, audio cue is played to indicate that
- aria alert dialogue pops up with the content of the suggestion

I am not sure why you don't hear the audio cue getting played. When you run Help: List Audio Cues and go down to the inline suggestion one, does it play?

**Comment by meganrogge** [comment]
To be clear, the suggestions impacted by this PR are only the ones that come from copilot. So instead of the popup saying "Suggestion: text to insert", it will say "text to insert"

**Comment by meganrogge** [comment]
@jooyoungseo I now understand the problem. 

We have 2 kinds of suggestions: inline and ones coming from a widget.

This PR deals only with the inline suggestions - the ones that appear immediately for sighted users. The widget suggestions have to be triggered to become visible. 

I agree with you that we should play an audio cue for either of the cases. But we probably will want to have a separate sound because to interact with widget suggestions, a different keybinding has to be used.

**Comment by jooyoungseo** [comment]
@meganrogge -- I agree with your ideas using two different audio cues to indicate the suggestion types. The one that I wanted to keep consistent is the pair of audioCue-aria-live event for either of suggestions.

**Comment by meganrogge** [comment]
I am not sure I understand, but think that you're suggesting the other suggestion type - those coming from the widget, should also have aria-live assertive alert applied. 

If so, I can look into that when adding the audio cue for that type of suggestion.

File: src/vs/editor/common/config/editorOptions.ts [hunk 1]
```diff
@@ -4929,6 +4935,13 @@ export const EditorOptions = {
 	ariaLabel: register(new EditorStringOption(
 		EditorOption.ariaLabel, 'ariaLabel', nls.localize('editorViewAccessibleLabel', "Editor content")
 	)),
+	ariaAssertiveInlineSuggestion: register(new EditorBooleanOption(
+		EditorOption.ariaAssertiveInlineSuggestion, 'ariaAssertiveInlineSuggestion', false,
```
Comment (isidorn) [inline src/vs/editor/common/config/editorOptions.ts:4939]: aria assertive sounds like an implementation detail. Can we have a more user friendly name?

File: src/vs/editor/contrib/inlineCompletions/browser/ghostTextWidget.ts [hunk 2]
```diff
@@ -274,6 +275,10 @@ class DecorationsWidget implements IDisposable {
 		// Using change decorations ensures that we update the id's before some event handler is called.
 		this.editor.changeDecorations(accessor => {
 			this.decorationIds = accessor.deltaDecorations(this.decorationIds, parts.map<IModelDeltaDecoration>(p => {
+				if (this.editor.getOption(EditorOption.ariaAssertiveInlineSuggestion)) {
+					this._alertWithContent(withNullAsUndefined(document.querySelector('.ghost-text-decoration')?.textContent));
```
Comment (isidorn) [inline src/vs/editor/contrib/inlineCompletions/browser/ghostTextWidget.ts:279]: Is this the best way to get the current inline completions? Can we somehow get it from the model, and not from HTML?

**Review by isidorn — DISMISSED** [review]

File: src/vs/editor/contrib/inlineCompletions/browser/ghostTextWidget.ts (hunk 2 above)
Comment (hediet) [reply src/vs/editor/contrib/inlineCompletions/browser/ghostTextWidget.ts:279]: Notice that the suggested text might have holes, e.g. in an if statement:
```
if (|[x === 1])[ { }]
```

The cursor is at `|`, everything in `[`...`]` is a suggestion.

You should probably use `parts` to figure out what to read.

**Review by hediet — CHANGES_REQUESTED** [review]
See comment

**Comment by meganrogge** [comment]
something I just noticed is now the aria alert happens before the audio cue is played, which is not good. working on the fix for that

**Comment by jooyoungseo** [comment]
@meganrogge -- I hear you because I encountered the same issue when working on my own project. Could we make the audio cue more responsive and played earlier? We want to keep the aria-live as responsive as possible without any silent filler to unnecessarily wait for the audio cue trigger.

**Comment by meganrogge** [comment]
I have it working nicely now, but how I did that was a bit hacky. It's a tricky problem since the code is synchronous and needs to remain that way.

File: src/vs/editor/contrib/inlineCompletions/browser/ghostTextWidget.ts [hunk 3]
```diff
@@ -74,6 +77,9 @@ export class GhostTextWidget extends Disposable {
 	private readonly replacementDecoration = this._register(new DisposableDecorations(this.editor));
 
 	private update(): void {
+		if (this._ariaStringBuilder.build().length) {
+			return;
+		}
```
Comment (hediet) [inline src/vs/editor/contrib/inlineCompletions/browser/ghostTextWidget.ts:82]: Why is this private state? I think it makes more sense to allocate this per update call (in a local variable).

**Comment by hediet** [comment]
Afaik the ghost text class has some methods to turn it into text, that is also used for the hover.

File: src/vs/editor/contrib/inlineCompletions/browser/ghostTextWidget.ts (hunk 3 above)
Comment (meganrogge) [reply src/vs/editor/contrib/inlineCompletions/browser/ghostTextWidget.ts:82]: Yep, makes sense

**Comment by hediet** [comment]
This deprecates AudioCue.inlineSuggestion, right?

**Review by hediet — APPROVED** [review]

**Comment by meganrogge** [comment]
@hediet maybe i'm misunderstanding, but that's still used here

https://github.com/microsoft/vscode/pull/174606/files#diff-7fef961bb90ba02a1ed54aeb24180e126845a809d8a90360a3d5e2c8935050b8R163

**Comment by hediet** [comment]
I think now the audio cue for inline suggestions no longer debounces and doesn't trigger on change anymore. I.e. when you cycle between suggestions I think it plays the sound now every time.

**Comment by meganrogge** [comment]
I think that might be desirable, but we can seek feedback. @jooyoungseo can you pls try this out?

**Comment by meganrogge** [comment]
@hediet based on the user's testing, this wasn't effective https://github.com/microsoft/vscode/issues/175105. should we revert it?

**Comment by hediet** [comment]
Then we should revert I think!

**Pull request merged**